### PR TITLE
Help should print required arguments first (#325)

### PR DESCRIPTION
**Pull request by cmnbroad**
Prints required args first follow by optional args, if any.

### CODE REVIEW

**Comment by akiezun** [comment]
Please add tests for this change

On Sunday, June 14, 2015, cmnbroad <notifications@github.com> wrote:

> Prints required args first follow by optional args, if any.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/broadinstitute/hellbender/pull/566
> Commit Summary
>
>    - Help should print required arguments first (#325)
>
> File Changes
>
>    - *M*
>    src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java
>    <https://github.com/broadinstitute/hellbender/pull/566/files#diff-0>
>    (18)
>
> Patch Links:
>
>    - https://github.com/broadinstitute/hellbender/pull/566.patch
>    - https://github.com/broadinstitute/hellbender/pull/566.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/hellbender/pull/566>.
>


-- 
Sent from Gmail Mobile

**Comment by akiezun** [comment]
@lbergelson please review

File: src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java [hunk 1]
```diff
@@ -181,11 +182,24 @@ public String getVersion() {
     public void usage(final PrintStream stream, final boolean printCommon) {
         stream.print(getStandardUsagePreamble(callerArguments.getClass()) + getUsagePreamble());
         stream.println("\n" + getVersion());
-        stream.println("\n\nArguments:\n");
+        stream.println("\n\nRequired Arguments:\n");
+
+        // print required args first; temporarily buffer output for optional args
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java]: This business with a second output stream seems overly clever.  I think just explicitly printing one and then the other would be much easier to read.

Something like this:
```
   List<ArgumentDefinition> args = argumentDefinitions.stream()
                .filter(argumentDefinition -> printCommon || !argumentDefinition.isCommon)
                .collect(Collectors.toList());

      stream.println("\n\nRequired Arguments:\n");
        args.stream().filter(a -> !a.optional).forEachOrdered(a -> printArgumentUsage(stream, a));

        stream.println("Optional Arguments");
        args.stream().filter(a -> a.optional).forEachOrdered(a -> printArgumentUsage(stream, a));
```

**Comment by lbergelson** [comment]
@cmnbroad back to you.  If you can come up with a meaningful test for this it wouldn't be a bad thing as well.

**Comment by cmnbroad** [comment]
@lbergelson You're right, it would be easier to read that way, but it leaves a dangling "Optional Arguments" string in the output even when there are no optional arguments. If you're still not sold I can rewrite it to iterate once to count the optional args, but this is a cheap, simple way to get the right output.

**Comment by cmnbroad** [comment]
Will add a test...

**Comment by lbergelson** [comment]
Oh, good point about the dangling Optional.  Likewise you could potentially have a tool with no required arguments.

**Comment by lbergelson** [comment]
I think I still prefer the version without the extra buffer stream, even if you have to explicitly count the arguments.  

Your way is cooler, but it made me stop for a second to understand what was happening.  The argument system already is convoluted enough, I think we should err on the side of "dumb but readable."

**Comment by akiezun** [comment]
yes, process multiple times if it makes the code more readable. speed is not an issue here at all.

**Pull request closed without merging**